confidant/services/jwkmanager.py

```diff
@@ -0,0 +1,109 @@
+import jwt
+import json
+
+from jwcrypto import jwk
+from typing import Any, Dict, Optional
+from hashlib import sha1
+from cryptography.x509 import load_pem_x509_certificate
+from cryptography.hazmat.primitives import serialization
+from confidant.settings import CERTIFICATE_AUTHORITIES, \
+    DEFAULT_JWT_EXPIRATION_SECONDS
+from datetime import datetime, timezone, timedelta
+
+
+class JWKManager:
+    def __init__(self) -> None:
+        self._keys = jwk.JWKSet()
+        self._public_keys = {}
+        self._token_cache = {}
+        self._payload_cache = {}
+
+        self._load_certificate_authorities()
+
+    def _load_certificate_authorities(self) -> None:
+        if CERTIFICATE_AUTHORITIES:
+            for ca in CERTIFICATE_AUTHORITIES:
+                self.set_key(ca['name'], ca['key'],
+                             passphrase=ca['passphrase'])
+
+    def set_key(self, kid: str, private_key: str,
+                passphrase: Optional[str] = None,
+                encoding: str = 'utf-8') -> str:
+        if passphrase:
+            passphrase = passphrase.encode(encoding)
+        key = jwk.JWK()
+        key.import_from_pem(private_key.encode(encoding),
+                            password=passphrase, kid=kid)
+        if not self._keys.get_key(kid):
+            self._keys.add(key)
+        return kid
+
+    def get_jwt(self, kid: str, payload: dict,
+                expiration_seconds: int = DEFAULT_JWT_EXPIRATION_SECONDS,
+                algorithm: str = 'RS256') -> str:
+        key = self._keys.get_key(kid)
+        if not key:
+            raise ValueError('This private key is not stored!')
+
+        now = datetime.now(tz=timezone.utc)
+        payload.update({
+            'iat': now,
+            'nbf': now,
+            'exp': now + timedelta(seconds=expiration_seconds),
+        })
+
+        payload_hash = sha1(
```

Comment:
I think this is going to be expensive here, calculating sha hashes on each jwt fetch.  I'm thinking we can use the downstream service role name only as the "key".  Then the values includes the expiry and the JWT.  

Does the payload currently include the requesting service role name?

Reply:
Actually I think this part doesn't make a lot of sense any more (this was here before adding the expiry field). The purpose of this was to cache tokens so we don't generate on each call. Do we want to keep caching tokens for X amount of time? (e.g. send back the same token for its lifespan?)

Reply:
I feel having a lever to either disable or enable it would be a good idea.  

It's unclear to me the CPU intensity to generate the tokens at moment (cryptographic functions usually have higher CPU requirements), but allowing us to easily enable/disable will allow us to experiment and enable us to keep CPU down (at expense of some memory).

Reply:
Added, the only downside I noticed is that caching this way will cause up to (# of replicas * # of workers) tokens at the same time for the same user, unless we support a distributed caching tool (since right now each worker has its own cache in memory)